datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -867,29 +908,56 @@ public void cleanupDeadInstanceAssignments(List<String> instances) {
         _liveTaskMap.remove(instance);
       }
     }
+    LOG.debug("unusedTasks remaining after dead instances cleanup: {}", unusedTasks.stream()
+        .map(DatastreamTask::getDatastreamTaskName)
+        .collect(Collectors.toList()));
+    // Clean the remaining unusedTasks.
+    cleanUpOldUnusedTasksFromConnector(unusedTasks);
+    LOG.info("cleanUpDeadInstanceDataAndOtherUnusedTasks completed successfully.");
   }
 
   /**
+   * It finds old unused tasks.
+   *
    * New assignment may not contain all the tasks from the previous assignment. This means that the diff of the
    * tasks between the new and old assignment are not used any more which can be deleted.
    * @param previousAssignmentByInstance Previous task assignment
    * @param newAssignmentsByInstance New task assignment.
+   * @return Tasks absent in newAssignment but present in oldAssignment.
    */
-  public void cleanupOldUnusedTasks(Map<String, Set<DatastreamTask>> previousAssignmentByInstance,
+  private Set<DatastreamTask> findOldUnusedTasks(Map<String, Set<DatastreamTask>> previousAssignmentByInstance,
       Map<String, List<DatastreamTask>> newAssignmentsByInstance) {
 
     Set<DatastreamTask> newTasks =
         newAssignmentsByInstance.values().stream().flatMap(Collection::stream).collect(Collectors.toSet());
     Set<DatastreamTask> oldTasks =
         previousAssignmentByInstance.values().stream().flatMap(Collection::stream).collect(Collectors.toSet());
-    List<DatastreamTask> unusedTasks =
-        oldTasks.stream().filter(x -> !newTasks.contains(x)).collect(Collectors.toList());
-    LOG.info("Deleting the unused tasks {} ", unusedTasks);
+    Set<DatastreamTask> unusedTasks =
+        oldTasks.stream().filter(x -> !newTasks.contains(x)).collect(Collectors.toSet());
+    LOG.info("Unused tasks to be deleted: {}", unusedTasks);
+    return unusedTasks;
+  }
 
+  /**
+   * Delete the unused connector tasks from zookeeper.
+   * @param unusedTasks Unused tasks.
+   */
+  private void cleanUpOldUnusedTasksFromConnector(Collection<DatastreamTask> unusedTasks) {
```

Comment:
Can you add logs for when this starts and when this completes? Want to be able to figure out how long this process takes

Reply:
There are start and end logs in the api calling this api which will take care of it. Also, the api inside this has api for each task.

Reply:
Adding logs here may be useful for other callers of this function that aren't going through cleanUpDeadInstanceDataAndOtherUnusedTasks()

Reply:
"LOG.info("Unused tasks to be deleted: {}", unusedTasks);" This is the log present in the other path as well to detect.

Reply:
And is end time also logged for other code paths?

Reply:
What if I want to find out how long these take to complete without it getting interrupted? I'm just trying to get data on how long these cleanup phases take on average. Today we only know when each event in the coordinator starts and ends. There is no way to track how long cleanup itself takes. Knowing this may be important to make a call on whether we should minimize interruptions during cleanup or not in the future.

Reply:
Added the log that will cover both the conversations.